rust: add `to_str` and `as_str_unchecked` to `CStr`

mdaverde opened the pull request:
Following the `std::ffi::CStr` pattern of `to_str`, this adds
methods to convert between CStr and &str both safely and unsafely

Closes #532 

Signed-off-by: Milan Landaverde <milan@mdaverde.com>

ojeda commented on rust/kernel/str.rs:
```diff
@@ -167,6 +167,50 @@ impl CStr {
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
```
Since there is only a single example, I would avoid it:

```suggestion
```

ojeda commented on rust/kernel/str.rs:
```diff
@@ -167,6 +167,50 @@ impl CStr {
+    /// # use kernel::c_str;
+    /// # use kernel::str::CStr;
+    /// // SAFETY: This is safe because string literals are guaranteed to be
+    /// // valid UTF-8 by the Rust compiler.
```
Perhaps simplify here too, i.e. `SAFETY` comments are always about justifying why it is safe:

```suggestion
    /// // SAFETY: String literals are guaranteed to be valid UTF-8
    /// // by the Rust compiler.
```

ojeda commented on rust/kernel/str.rs:
```diff
@@ -167,6 +167,50 @@ impl CStr {
+    ///
+    /// Callers *must* ensure that CStr is valid UTF-8 before
+    /// calling this method. Converting a CStr into a `&str` that is
+    /// not valid UTF-8 is considered undefined behavior.
```
I would simplify:

```suggestion
    /// The contents must be valid UTF-8.
```

ojeda commented on rust/kernel/str.rs:
```diff
@@ -251,3 +295,32 @@ macro_rules! c_str {
+    }
+
+    #[test]
+    fn test_cstr_as_str() {
```
```suggestion
    fn test_cstr_as_str_unchecked() {
```

mdaverde replied:
Updated!

ojeda commented on rust/kernel/str.rs:
```diff
@@ -251,3 +295,32 @@ macro_rules! c_str {
+    fn test_cstr_to_str_panic() {
+        let bad_bytes = b"\xc3\x28\0";
+        let unchecked_c_str = unsafe { CStr::from_bytes_with_nul_unchecked(bad_bytes) };
+        unchecked_c_str.to_str().unwrap();
```
Why does this need to be unchecked? i.e. `CStr` admits any sequence, UTF-8 or not.

ojeda commented:
A couple more nits:

  - Please add intra-doc links wherever possible, e.g. in mentions to `CStr` and `&str`.
  - In the commit message: the functions are about converting `CStr` to `&str`, not "between" (this does not matter much today, since the commit messages do not go into the kernel, but eventually they will! :)

gurugio commented:
Hi,

I would like to run the unit-tests in your patch.
Could you please inform me how to run the unit-tests in the patch?
I knew how to run the unit-test in normal user-mode application with Rust
but I've not seen the unit-test in kernel code.

I have a VM running the kernel with Rust and loading the modules implemented with Rust.
I ran the user-mode application unit-test with VisualStudio and Cargo in the VM.
I guess it would be very different to run the unittest of kernel code.

mdaverde commented:
From @ojeda 
> A couple more nits:
> 
>     * Please add intra-doc links wherever possible, e.g. in mentions to `CStr` and `&str`.
Went ahead and added these

> 
>     * In the commit message: the functions are about converting `CStr` to `&str`, not "between" (this does not matter much today, since the commit messages do not go into the kernel, but eventually they will! :)
Updated this commit message as well

mdaverde commented:
From @gurugio 
> I would like to run the unit-tests in your patch. Could you please inform me how to run the unit-tests in the patch? I knew how to run the unit-test in normal user-mode application with Rust but I've not seen the unit-test in kernel code.

Hello! I've been running the tests with `make LLVM=1 rusttest`

ojeda commented:
Looks good to me! Thanks for contributing!
